exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/statistics/StatisticsAggBatch.java

```diff
@@ -0,0 +1,256 @@
+import org.apache.drill.exec.physical.impl.aggregate.StreamingAggTemplate;
+import org.apache.drill.exec.physical.impl.aggregate.StreamingAggregator;
+import org.apache.drill.exec.planner.physical.StatsAggPrel.OperatorPhase;
+import org.apache.drill.exec.record.BatchSchema.SelectionVectorMode;
+import org.apache.drill.exec.record.MaterializedField;
+import org.apache.drill.exec.record.RecordBatch;
+import org.apache.drill.exec.record.TypedFieldId;
+import org.apache.drill.exec.store.ImplicitColumnExplorer;
+import org.apache.drill.exec.vector.ValueVector;
+import org.apache.drill.exec.vector.complex.FieldIdUtil;
+import org.apache.drill.exec.vector.complex.MapVector;
+
+import java.io.IOException;
+import java.util.GregorianCalendar;
+import java.util.List;
+import java.util.TimeZone;
+
+/**
+ * TODO: This needs cleanup. Currently the key values are constants and we compare the constants
+ * for every record. Seems unnecessary.
+ *
+ * Example input and output:
+ * Schema of incoming batch: region_id (VARCHAR), sales_city (VARCHAR), cnt (BIGINT)
+ * Schema of output:
+ *    "schema" : BIGINT - Schema number. For each schema change this number is incremented.
+ *    "computed" : BIGINT - What time is it computed?
+ *    "columns"       : MAP - Column names
+ *       "region_id"  : VARCHAR
+ *       "sales_city" : VARCHAR
+ *       "cnt"        : VARCHAR
+ *    "statscount" : MAP
+ *       "region_id"  : BIGINT - statscount(region_id) - aggregation over all values of region_id
+ *                      in incoming batch
+ *       "sales_city" : BIGINT - statscount(sales_city)
+ *       "cnt"        : BIGINT - statscount(cnt)
+ *    "nonnullstatcount" : MAP
+ *       "region_id"  : BIGINT - nonnullstatcount(region_id)
+ *       "sales_city" : BIGINT - nonnullstatcount(sales_city)
+ *       "cnt"        : BIGINT - nonnullstatcount(cnt)
+ *   .... another map for next stats function ....
+ */
+public class StatisticsAggBatch extends StreamingAggBatch {
+  private List<String> functions;
+  private int schema = 0;
+
+  public StatisticsAggBatch(StatisticsAggregate popConfig, RecordBatch incoming,
+      FragmentContext context) throws OutOfMemoryException {
+    super(popConfig, incoming, context);
+    this.functions = popConfig.getFunctions();
+  }
+
+  private void createKeyColumn(String name, LogicalExpression expr, List<LogicalExpression> keyExprs,
+      List<TypedFieldId> keyOutputIds) throws SchemaChangeException {
+    ErrorCollector collector = new ErrorCollectorImpl();
+
+    LogicalExpression mle = ExpressionTreeMaterializer.materialize(expr, incoming, collector,
+        context.getFunctionRegistry());
+
+    MaterializedField outputField = MaterializedField.create(name, mle.getMajorType());
+    ValueVector vector = TypeHelper.getNewVector(outputField, oContext.getAllocator());
+
+    keyExprs.add(mle);
+    keyOutputIds.add(container.add(vector));
+
+    if (collector.hasErrors()) {
```

Comment:
Moved it earlier as suggested. However, the ErrorCollector collects ExpressionValidationError which is not subclassed from error or exception classes. We would need to subclass it and generate the equivalent error/exception in order to re-throw it.